$p->resolve($p) rejects with an error message, and only one call to resolve considered

akarelas opened the pull request:
### Summary
Sort of solves bug #1517 (`$p->resolve($p)` should reject an exception). I say "sort of" because the reason of the rejection is not a TypeError object (itself a subclass of Error), but a simple text message, because I don't know how to produce errors (and if they are at all relevant to Perl). Nevertheless, with this PR I provide a solution the maintainers can modify to make an Error if they so wish.

**Edit:** This PR now also solves bug #1515.

### Motivation
This change makes Mojo::Promise objects more in accordance with the spec (https://tc39.es/ecma262/#sec-promise-resolve-functions)

### References
* Bug #1517 
* https://tc39.es/ecma262/#sec-promise-resolve-functions point 7
* https://tc39.es/ecma262/#sec-promise-resolve-functions points 5 & 6

jhthorsen commented on lib/Mojo/Promise.pm:
```diff
@@ -215,7 +215,8 @@ sub _settle {
   if ($thenable && $status eq 'resolve') {
-    $results[0]->then(sub { $self->resolve(@_); () }, sub { $self->reject(@_); () });
+    if (refaddr $results[0] == refaddr $self) {$self->reject('Chaining cycle detected')}
+    else { $results[0]->then(sub { $self->resolve(@_); () }, sub { $self->reject(@_); () }); }
```
I don't think this is run through perltidy. I suggest this formatting:

```
    (refaddr $results[0] == refaddr $self)
      ? $self->reject('Chaining cycle detected.')
      : $results[0]->then(sub { $self->resolve(@_); () }, sub { $self->reject(@_); () });
```

akarelas replied:
I don't have access to perltidy right now. You'd have to run it through perltidy yourself, if that's not a huge problem for you. Thanks.

akarelas replied:
addressed

jhthorsen reviewed (changes requested):
This PR cannot be accepted without any accommodating tests.

I'm unsure if this is a real issue though or simply an academic one, but I'm positive to the change.

christopherraa commented:
I'm also positive to the change, but there needs to be tests. Please add tests for this change and I'll follow up by reviewing it.

akarelas commented:
This PR now solves two issues (#1515 and #1517), contains tests, and have passed everything (`lib/Mojo/Promise.pm` and `t/mojo/promise.t`) through perltidy.

kraih commented on lib/Mojo/Promise.pm:
```diff
@@ -214,8 +214,19 @@ sub _settle {
     $self = $self->new;
   }
 
+  !$self->{_semi_settled}++ or return $self;
```
We don'y use underscore names.

kraih replied:
Or this `... or return ...` style.

akarelas replied:
Addressed.

kraih commented on t/mojo/promise.t:
```diff
@@ -516,4 +516,29 @@ subtest 'Wait for stopped loop' => sub {
 
+subtest 'Resolve self' => sub {
+  my $promise = Mojo::Promise->new;
+  my @results = ([], []);
```
Other tests don't use `@results` for both.

akarelas replied:
Addressed.

kraih commented on t/mojo/promise.t:
```diff
@@ -516,4 +516,29 @@ subtest 'Wait for stopped loop' => sub {
+  my @results = ([], []);
+  $promise->resolve($promise)->then(sub { $results[0] = \@_ }, sub { $results[1] = \@_ });
+  Mojo::IOLoop->one_tick;
+  is_deeply $results[0], [], 'no resolve results';
```
Inconsistent description.

akarelas replied:
Addressed.

kraih commented:
Your commit messages are really bad, to the point where i will have to reject the PR for it. And you did not squash the commits.

kraih commented on lib/Mojo/Promise.pm:
```diff
@@ -214,8 +214,19 @@ sub _settle {
-    $results[0]->then(sub { $self->resolve(@_); () }, sub { $self->reject(@_); () });
+    if (refaddr $results[0] == refaddr $self) {
+      undef $self->{_semi_settled};
+      $self->reject('Chaining cycle detected');
```
Isn't the JavaScript exception `Chaining cycle detected for promise`?

akarelas replied:
Addressed.

**node.js** exception is `Chaining cycle detected for promise [Promise]`. Remembering how **node.js** outputs `[object Object]` when you stringify an object (as in `console.log('hi: ' + {})`), it came to my mind that it would be more correct, and helpful to developers who are investigating their own bugs at the same time, if I stringified $self (a promise) in this message too (as in: `"Chaining cycle detected for promise $self"`).

If you don't agree with this I will remove $self from that string.

kraih commented on lib/Mojo/Promise.pm:
```diff
@@ -214,8 +214,19 @@ sub _settle {
-    $results[0]->then(sub { $self->resolve(@_); () }, sub { $self->reject(@_); () });
+    if (refaddr $results[0] == refaddr $self) {
+      undef $self->{semi_settled};
+      $self->reject("Chaining cycle detected for promise $self");
```
What is the use of including `$self` there?

akarelas replied:
The use is that this kind of cycling error is only going to happen in a bug. So I included `$self` here to allow the developer to debug and figure out which of their promises took part in the cycle. (I'm assuming the developer can insert debugging code in their own code to display the memory address of their promises, but shouldn't have to tweak the Mojolicious library's code to display the address of the promise that took part in the cycle).

akarelas replied:
Plus, node.js also displays `...for promise [Promise]` in their error message, which kind of reminds how objects are stringified (`[object Object]`), so perhaps including `$self` increases the consistency with node.js.

akarelas commented:
> Your commit messages are really bad, to the point where i will have to reject the PR for it. And you did not squash the commits.

Fixed messages and squashed commits, but now the message takes 2 lines.

kraih commented on t/mojo/promise.t:
```diff
@@ -516,4 +516,30 @@ subtest 'Wait for stopped loop' => sub {
+  $promise->resolve($promise)->then(sub { @results = @_ }, sub { @errors = @_ });
+  Mojo::IOLoop->one_tick;
+  is_deeply \@results, [], 'promise not resolved';
+  cmp_ok scalar(@errors), '==', 1, 'correct number of errors';
```
How is that better than `is_deeply \@errors, [], ...`?

kraih commented on t/mojo/promise.t:
```diff
@@ -516,4 +516,30 @@ subtest 'Wait for stopped loop' => sub {
+  like $errors[0], qr/^Chaining cycle detected for promise \S/, 'correct error';
+};
+
+subtest 'Resolve to unsettled promise' => sub {
```
Is it resolve "to" or resolve "with" unsettled promise?

akarelas replied:
I do not know. Couldn't find either of the two in past tests. I believe I haven't used both verbs in my commit.

But I believe "resolve to" conveys the message in a more clear way in this case, because "Resolve with unsettled promise" could also mean "calling `resolve` on an unsettled promise" which is not what I meant.

* This promise resolves to 5
* This promise resolves with 5

Maybe both expressions are valid, but I can't say. I'm not that good with English prepositions.

akarelas replied:
Since I saw you were already using the preposition "with" in similar -although non-ambiguous- situations inside `mojo/promise.t`, I decided to keep the same preposition, and addressed the ambiguity by renaming "Resove to unsettled promise" to "Settle with unsettled promise" and added the `$p->reject($p2)` case inside that subtest, too. Also I renamed "Resolve to self" to "Resolve with self", as that was already non-ambiguous.

To make sure the case "Reject with self" is covered, I renamed the subtest "Resolve to self" to "Settle with self" and added the "Reject with self" case in there too.

akarelas replied:
shadowcat-mst below provided the answer to your question: "resolve "to" or resolve "with"" (it is "resolve with"). I have thus renamed the subtests, and included more cases in them (with "reject" instead of "resolve").

Anything else you need to accept this PR?

akarelas commented:
I created this looping test script, which tests 10 settling cases for Promises. e.g. `Mojo::Promise->resolve($p);` and `$p->reject($p)`, etc.

Maybe you might want to keep a copy to test all 10 cases next time you change something in Mojo::Promise. I created it to test for bugs, and to check whether I fixed them in my PR.

https://gist.github.com/akarelas/1cddeda23e6a408c31929b3a1993964f

shadowcat-mst commented:
If you mean to resolve a promise that itself is unsettled,

  $unsettled->resolve(...)

would be "resolve unsettled promise"

If you mean to resolve a promise by passing another unsettled promise,

  $promise->resolve($unsettled)

would be "resolve with unsettled promise"

akarelas commented:
@christopherraa , @jhthorsen I have addressed all the issues you and kraih had with this bugfix.

Could you please look at it again and approve it, if it passes your criteria?

Thanks.

kraih commented:
I don't see this PR going anywhere. The complexity is just way too high for the (still questionable) value it would add.

akarelas commented:
> I don't see this PR going anywhere. The complexity is just way too high for the (still questionable) value it would add.

I'd like to remind that it also solves bug #1515. Anyway, thought it would be cool to solve two bugs with 15 lines of code. Your choice.

kraih commented on t/mojo/promise.t:
```diff
@@ -368,6 +368,47 @@ subtest 'Settle with promise' => sub {
+  Mojo::IOLoop->one_tick;
+  is_deeply \@results, [], 'promise not resolved';
+  is_deeply \@errors, [$promise2], 'correct errors';
+};
```
This block of tests seems completely unrelated to the PR description.

akarelas replied:
Indeed, this block was related to the second issue this PR solves, which was added later. (For history reasons I thought I shouldn't have modified any of the past comments, including the title & description). Now the PR's description has been updated to address that.

kraih replied:
PRs trying to resolve multiple issues at once pretty much never get accepted.

kraih commented on t/mojo/promise.t:
```diff
@@ -349,7 +349,7 @@ subtest 'All settled (with rejection)' => sub {
 };
 
-subtest 'Settle with promise' => sub {
+subtest 'Settle with settled promise' => sub {
```
These tests are unrelated to your PR, why would you randomly change their description?

akarelas replied:
I was under the impression that a patch needs to leave the repo in a consistent state. It would seem strange to have a subtest that would test "Settle with promise", and another one that would test "Settle with unsettled promise". Because, what would the second subtest's reason for existence be, if its case is already being tested by the first subtest? So I did this modification.

Shall I revert that line?

kraih reviewed (changes requested):
I'll block this PR because it contains random unexplained changes.

kraih commented:
Has this PR been abandoned?

jhthorsen commented:
I cannot remember seeing this problem while writing code, so I'm no longer positive to the change. I consider the PR to have academic value, but not something that is frequent enough in real life to be considered.

akarelas commented:
> Has this PR been abandoned?

No, sorry, I was just busy celebrating Christmas. Now I'm back to working.

akarelas commented:
> I cannot remember seeing this problem while writing code, so I'm no longer positive to the change. I consider the PR to have academic value, but not something that is frequent enough in real life to be considered.

Indeed, the only reason I made this PR was to improve compliance with the ES spec. None of my code would encounter these issues either, unless my code had a bug. I'm okay with you not merging this PR.

kraih commented:
It seems nobody has ever actually run into these issues and the code complexity is fairly high. So i'm 👎 on this PR.

jberger reviewed (changes requested):
After letting this sit this long, I've got to agree. It is a lot of complexity on top of an already complex (of necessity) codebase. If it isn't addressing a real world need (like warning on unhandled errors) i vote no

kraih commented:
Thanks for your contribution, but i don't see this PR pass a vote.
